File: Amazon/Onsite/Next Greater Node In Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 
 Exactly same solution with Next Greater Element I, II, except for linked list we need to store both index & val
 Time Complexity : O(N)
 Space Complexity : O(N) ->worst case of stack depth = N(decresing array)
 
 */
class Solution {
public:
    vector<int> nextLargerNodes(ListNode* head) {
        stack<pair<int,int>> stk; //store index
        vector<int> res;
        ListNode* curr = head;
        int i = 0;
        while(curr!=NULL)
        {
            while(!stk.empty() && stk.top().second < curr->val) {
                int idx = stk.top().first; stk.pop();
                res[idx] = curr->val;
            }
            stk.emplace(i, curr->val);
            res.push_back(0);
            i++;
            curr = curr->next;
        }
        
        return res;
    }
};

File: NextGreaterNodeinLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def nextLargerNodes(self, head: ListNode) -> List[int]:
        # convert linked list to array, user extra space
        nums = []
        cur = head
        while cur:
            nums.append(cur.val)
            cur = cur.next
        # monotonic decreasing stack, same as 
        res = []
        stack = []
        n = len(nums)
        for i in range(n-1, -1, -1):
            while stack and stack[-1] <= nums[i]:
                stack.pop()
            if len(stack) == 0:
                res.append(0)
            else:
                res.append(stack[-1])
            stack.append(nums[i])
        return res[::-1]
                